text/0000-futures.md

```diff
@@ -0,0 +1,617 @@
+    blocks some *other* desired impls (around `Box` and `&mut` specifically),
+    but the problem doesn't appear to be fundamental.
+
+  - Via a subtrait relationship, so that `T: Future` is defined essentially as
+    an alias for `for<'a> Pin<&mut 'a T>: MoveFuture`. Unfortunately, such
+    "higher ranked" trait relationships don't currently work well in the trait
+    system, and this approach also makes things more convoluted when
+    implementing `Future` by hand, for relatively little gain.
+
+The drawback of the "one core trait" approach taken by this RFC is its ergonomic
+hit when writing moveable futures by hand: you now need to import `Pin` and
+`Unpin`, invoke `poll_unpin`, and impl `Unpin` for your types. This is all
+pretty mechanical, but it's a pain. It's possible that improvements in `Pin`
+ergonomics will obviate some of these issues, but there are a lot of open
+questions there still.
+
+On the other hand, a two-trait approach has downsides as well. If we *also*
+remove the error type, there's a combinatorial explosion, since we end up
+needing `Try` variants of each trait (and this extends to related traits, like
+`Stream`, as well). More broadly, with the one-trait approach, `Unpin` acts as a
+kind of "independent knob" that can be applied orthogonally from other concerns;
+with the two-trait approach, it's "mixed in". And both of the two-trait
+approaches run up against compiler limitations at the moment, though of course
+that shouldn't be taken as a deciding factor.
+
+**The primary reason this RFC opts for the one-trait approach is that it's the
+conservative, forward-compatible option, and has proven itself in practice**.
+It's possible to add `MoveFuture`, together with a blanket impl, at any point in the future.
+Thus, starting with just the single `Future` trait as proposed in this RFC keeps our options
+maximally open while we gain experience.
+
+## Rationale, drawbacks and alternatives to the wakeup design (`Waker`)
+
+Previous iterations of this proposal included a separate wakeup type,
+`LocalWaker`, which was `!Send + !Sync` and could be used to implement
+optimized executor behavior without requiring atomic reference counting
+or atomic wakeups. However, in practice, these same optimizations are
+available through the use of thread-local wakeup queues, carrying IDs
+rather than pointers to wakeup objects, and tracking an executor ID or
+thread ID to perform a runtime assertion that a `Waker` wasn't sent across
+threads. For a simple example, a single thread-locked executor with zero
+atomics can be implemented as follows:
+
+```rust
+struct Executor {
+    // map from task id (usize) to task
+    tasks: Slab<Task>,
+    // list of woken tasks to poll
+    work_queue: VecDeque<usize>,
+}
+
+thread_local! {
+  pub static EXECUTOR: RefCell<Option<Executor>> = ...;
+}
+
+static VTABLE: &RawWakerVTable = &RawWakerVTable {
+    clone: |data: *const ()| RawWaker { data, vtable: VTABLE, },
```

Comment:
ugh really? I should just go add that coercion.

Reply:
Just adding the ability to coerce from closures to `unsafe fn`.

Reply:
Ah gotcha; that seems reasonable. Moreover, it should work if and when we add transitive coercions because we can just remove the closure -> unsafe fn coercion and do it transitively instead.

Reply:
I made a dedicated issue: https://github.com/rust-lang/rust/issues/57883